feat: Data set and USS download commands

JWaters02 opened the pull request:
## Proposed changes

<!-- Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. -->

Adds 5 new commands:
1. Download member
2. Download data set
3. Download all members
4. Download USS file
5. Download USS directory

Didn't add command for download all data sets even though its in the API because of [#2536](https://github.com/zowe/zowe-cli/issues/2536) but happy to add it if something thinks it would be useful to have it in ZE... 

https://github.com/user-attachments/assets/532628cd-5770-4f05-b8f7-0362c5bce2fa

## Release Notes

<!-- Include the Milestone Number and a small description of your change that will be added to the changelog. -->
<!-- If there is a linked issue, it should have the same milestone as this PR. -->

Milestone: 3.4.0

Changelog:

## Types of changes

<!-- What types of changes does your code introduce to Zowe Explorer? Put an `x` in all boxes that apply. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

<!-- Put an `x` in all boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer. -->

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] All PR dependencies have been merged and published (if applicable)
- [x] A GIF or screenshot is included in the PR for visual changes
- [ ] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`

**Code coverage**

- [ ] There is coverage for the code that I have added
- [ ] I have added new test cases and they are passing
- [ ] I have manually tested the changes

**Deployment**

- [ ] I have tested new functionality with the FTP extension and profile verifying no extender profile type breakages introduced
- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

## Further comments

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did, what alternatives you considered, etc... -->

Please could someone (ideally someone on zowe cli team who worked on downloads api) have a look at the options I am passing into downloads and if I am missing any/getting some wrong etc. Thanks

JWaters02 commented:
Should we have USS directory download allow user to choose the default encoding method for USS files being downloaded? I see in the API that if none is provided, each USS file will be downloaded with an estimate if it has tags. This might be best as if user is downloading big directories of files, USS files might be using all sorts of encoding. 

On the other hand, I suppose they could choose encoding, but the prompt could have an extra selection for "don't use any encoding" for directories so the user could both choose an encoding for all files or choose not to have any chosen by them/their profile. But this would be tricky to add as the prompt for encoding is used by upload and other download functions. Plus the current prompt for encoding does not support nodes being directories right now (I'd need to figure out something...).

Any thoughts?

anaxceron commented on packages/zowe-explorer/l10n/poeditor.json:
```diff
@@ -741,6 +732,15 @@
   "initializeUSSFavorites.error.buttonRemove": "",
   "File does not exist. It may have been deleted.": "",
   "Pulling from Mainframe...": "",
+  "The 'move' function is not implemented for this USS API.": "",
```
```suggestion
  "The 'Move' function is not implemented for this USS API.": "",
```

davenice commented:
I haven't tried the UX myself yet, but from the video I wasn't quite sure what I expected to happen about choosing encodings. The wording suggests I can choose the encoding of every file somehow ... but I can't quite envisage the GUI for this!

A couple of suggestions on the notifications....

I feel like when I get a few notifications that are quite generic like "File downloaded" or "New copy completed" and I find myself wondering if that was a notification I got just now, or one I got a while back that should have cleared. My stance for CICS has been to try and be a bit more descriptive with what the notification is e.g. "New copy completed on MYPROG1" or similar. We could be more descriptive in the notifications here, although I'm not set on that.

We could also provide a link in the notification to the file, or the directory we downloaded to. Again I don't think this is a must-have if there is a time pressure.

JWaters02 commented:
Thanks a lot for the feedback dave, I really appreciate it🙂 

The video is actually super out of date now I really should post a new one for better feedback, apologies!

> I haven't tried the UX myself yet, but from the video I wasn't quite sure what I expected to happen about choosing encodings. The wording suggests I can choose the encoding of every file somehow ... but I can't quite envisage the GUI for this!

Instead of putting encoding options right into the download options quick pick, I have an option for "Choose Encoding" and if it is checked, it will also enter the existing choose encoding quick pick used for upload/open with encoding.

For downloading members of a PDS, this encoding will be applied to all members, which does make sense given they all follow the same characteristics - but for a USS directory, files in a directory can have all kinds of encodings so this is much tricker. So for downloading USS directories, I have the same "Choose Encoding" option but the resulting encoding selection quick pick is almost the same as the existing one but with an additional "Auto-detect file encoding" option which essentialls selects no encoding and relies on the encoding detection in zowe-cli which uses USS file tags and stuff.

> A couple of suggestions on the notifications....
>
> I feel like when I get a few notifications that are quite generic like "File downloaded" or "New copy completed" and I find myself wondering if that was a notification I got just now, or one I got a while back that should have cleared. My stance for CICS has been to try and be a bit more descriptive with what the notification is e.g. "New copy completed on MYPROG1" or similar. We could be more descriptive in the notifications here, although I'm not set on that.

This is a good point, I think I will implement that. Currently, I have it only saying what type of download has occured but I definately should include the path of the thing downloaded and/or resulting file path too.

What I do have now though is when a warning or error occurs, the user can click "View Details" and it will open the Troubleshooting webview with the command response from zowe-cli which contains a lot more details about the download. These types of details can't be shown in notifcations because of being too verbose. A warning can occur, for example, if overwrite is off and a file was skipped because of the overwrite and in the details it prints which files were skipped and not skipped (same as in zowe-cli (whch I have an open PR for to add overwrite support to all commands except USS directories (which it already exists for), so you can currently only test this with those)).

> We could also provide a link in the notification to the file, or the directory we downloaded to. Again I don't think this is a must-have if there is a time pressure.

A link is a good idea!

traeok commented:
> I added a new test file USSUtils.unit.test.ts but when I run it, I get this error:
> 
> <snip>
> Unable to import package.json to get app Options : TypeError [ERR_INVALID_ARG_TYPE]: The "paths[0]" argument must be of 
> type string. Received undefined

Still debugging this, it's a strange error with the Jest JUnit reporting where it fails to locate the package JSON. Plan to read over some of the changes to get a better grasp of why this is happening

traeok commented:
> Only vague thought I had is that I tried to use MockedProperty as much as possible instead of Object.defineProperty and maybe that usage has caused some weird side effect?

Skimming through the tests, I do see a few `new MockedProperty` calls that are not calling the `Symbol.dispose` function, that could have an impact depending on what's being mocked

t1m0thyj commented on packages/zowe-explorer/src/trees/uss/UssFSProvider.ts:
```diff
@@ -433,10 +433,17 @@ export class UssFSProvider extends BaseProvider implements vscode.FileSystemProv
+            return file.encoding;
+        } catch (error) {
+            if (error instanceof vscode.FileSystemError && error.code === "FileIsADirectory") {
+                throw new Error(vscode.l10n.t("Cannot fetch encoding for directories. Encoding options are only available for files."));
```
Curious why this extra error check is needed, since we were already returning a clear error code (`FileIsADirectory`) previously.

JWaters02 replied:
thanks mr fernando and mr nice!

looked at this again yes I think this was just confusion on my part, good spot

JWaters02 commented:
> Skimming through the tests, I do see a few `new MockedProperty` calls that are not calling the `Symbol.dispose` function, that could have an impact depending on what's being mocked

Hey @traeok so I tried a few things:
1. Adding proper disposals to MockedProperty (in all places in test files were I was using it). Had no change
2. Replacing MockedProperty with Object.defineProperty. Had no change. So it surely can't be anything to do with MockedProperty? (a thought at the time of writing)
3. Running `pnpm test:unit -- --testPathPattern="USSUtils.unit.test"` did not work, which rules out some dodginess with npx and jest.
4. Made new test file from scratch with simple assertions tests worked, but as soon as I added the global mocks function (without even any tests using it), the error re-appeared on that file.
5. Made new test file and this time did not include any global mocks function, just doing all mocks in each test suite. I added a few tests at a time, all none without doing any MockedProperty/Object.defineProperty and all the tests were running WITHOUT the error from appearing. I got to the point where I had rewritten 23 of 23 test cases in my temp file without using global mocks and converting property mocks to spy ones in all but one test where I kept the mocked property on the constants mock (as I realised many property mocks didn't need to be).
6. So then I made a second test file which took my learnings and rewrote the original file again with global mocks to remove code duplication, but this time without mocking on built in node libraries and only mocking on stuff that really needed it. I also consulted chatgpt with original file and the first test file to ask it what it thought about the original file getting the error but the new one not, and its response was something about mocked property being instantiated before jest's mocking initialisation is done. I have no idea how any of that works so I asked it for a workaround for setting up the mocks properly and it seemed to work but again I don't understand this fact so it could be hallucinating and I wouldn't know. 

So off the back of points 5 and 6 I now replaced the original USSUtils with my temp one that is working so you can see what I eventually ended up with. Now I remember that when I was oriignally making the test file it was working up to a certain point when I started using the mocked properties inglobal mocks so thats also another factor I forgor about because it's been so long since I touched this,

JWaters02 commented on packages/zowe-explorer/l10n/bundle.l10n.json:
```diff
@@ -341,6 +305,42 @@
   "initializeUSSFavorites.error.buttonRemove": "initializeUSSFavorites.error.buttonRemove",
   "File does not exist. It may have been deleted.": "File does not exist. It may have been deleted.",
   "Pulling from Mainframe...": "Pulling from Mainframe...",
+  "The 'move' function is not implemented for this USS API.": "The 'move' function is not implemented for this USS API.",
```
refers to a function name which is lowercase so won't make it upper

awharn commented on packages/zowe-explorer-api/src/extend/MainframeInteraction.ts:
```diff
@@ -235,6 +250,16 @@ export namespace MainframeInteraction {
+         * Retrieve all members of a partitioned data set and save them to a directory.
+         *
+         * @param {string} dataSetName
+         * @param {string} directoryPath
```
This parameter doesn't seem to exist on `downloadAllMembers`

anaxceron commented on packages/zowe-explorer-api/CHANGELOG.md:
```diff
@@ -12,6 +12,8 @@ All notable changes to the "zowe-explorer-api" extension will be documented in t
 - Added support for loading credential manager options from the `imperative.json` file. Added a `credentialManagerOptions` object in the JSON object in `imperative.json` to specify options for the current credential manager. [#3935](https://github.com/zowe/zowe-explorer-vscode/pull/3935)
 - Added support for custom persistence levels for Windows (persist option) to support the credential manager in less permissive environments. For more information on how to configure this option, see the "Troubleshooting Zowe CLI credentials" on [Zowe Docs](https://docs.zowe.org/stable/troubleshoot/cli/troubleshoot-cli-credentials/#secrets-sdk-persistence-level-for-windows). [#3935](https://githu ...
+- Added support for `downloadDirectory` and `downloadAllMembers` to the `MainframeInteraction` namespace. [#3843](https://github.com/zowe/zowe-explorer-vscode/pull/3843)
+- Added support for calling `UssApi.fileList` with `zosfiles.IUSSListOptions`. [#3843](https://github.com/zowe/zowe-explorer-vscode/pull/3843)
```
We should describe what `UssApi.fileList` with `zosfiles.IUSSListOptions` are

anaxceron commented on packages/zowe-explorer-ftp-extension/CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ All notable changes to the "zowe-explorer-ftp-extension" extension will be docum
 ### New features and enhancements
 
 - Updated minimum VS Code version from 1.79 to 1.90. We are dropping support for VS Code releases that bundle versions of Node.js no longer receiving security updates. [#3684](https://github.com/zowe/zowe-explorer-vscode/issues/3684)
+- Added unimplemented functions for `downloadDirectory` and `downloadAllMembers`, that may be implemented in the future. [#3843](https://github.com/zowe/zowe-explorer-vscode/pull/3843)
```
We should describe what `downloadDirectory` and `downloadAllMembers` are

JWaters02 replied:
If I explain what they are, wouldn't it be confusing to the end user if the function is actually implemented or not? I only added this changelog entry to not fail the changelog check...

anaxceron commented on packages/zowe-explorer/CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 - Added support for loading credential manager options from the imperative.json file. Add a `credentialManagerOptions` object in the JSON object in `imperative.json` to specify options for the current credential manager. [#3935](https://github.com/zowe/zowe-explorer-vscode/pull/3935)
 - Added support for custom persistence levels for Windows (persist option) to support the credential manager in less permissive environments. For more information on how to configure this option, refer to the ["Troubleshooting Zowe CLI credentials"](https://docs.zowe.org/stable/troubleshoot/cli/troubleshoot-cli-credentials/#secrets-sdk-persistence-level-for-windows) page on Zowe Docs. [#3935](htt ...
 - Updated minimum VS Code version from 1.79 to 1.90. We are dropping support for VS Code releases that bundle versions of Node.js no longer receiving security updates. [#3684](https://github.com/zowe/zowe-explorer-vscode/issues/3684)
+- Added support for downloading data sets, data set members, USS files and USS directories. [#3843](https://github.com/zowe/zowe-explorer-vscode/pull/3843)
```
```suggestion
- Added support for downloading data sets, data set members, USS files, and USS directories. [#3843](https://github.com/zowe/zowe-explorer-vscode/pull/3843)
```

traeok commented on packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpUssApi.ts:
```diff
@@ -101,6 +101,23 @@ export class FtpUssApi extends AbstractFtpApi implements MainframeInteraction.IU
+     * @param listOptions Options for listing files in USS
+     * @returns A file response with the results of the download operation.
+     */
+    public downloadDirectory(
```
This implementation stub would not be necessary if the API was marked optional.

traeok commented on packages/zowe-explorer/src/trees/dataset/DatasetActions.ts:
```diff
@@ -618,6 +620,446 @@ export class DatasetActions {
+            fileExtension,
+        } = dataSetDownloadOptions;
+
+        await DatasetActions.executeDownloadWithProgress(
```
Yes, sorry for the late reply.

zFernand0 commented on packages/zowe-explorer-api/src/extend/MainframeInteraction.ts:
```diff
@@ -119,9 +119,24 @@ export namespace MainframeInteraction {
          *
          * @param {string} ussFilePath
          * @param {zosfiles.IDownloadOptions} options
+         * @returns {Promise<zosfiles.IZosFilesResponse>}
```
Thanks for adding the return value to the TSDoc. 🙏 
It would be nice if all functions were consistently returning the proper `Promise<...>` value 😅 
This is not a request for changes though 😋

zFernand0 commented on packages/zowe-explorer/src/trees/dataset/DatasetUtils.ts:
```diff
@@ -158,4 +158,58 @@ export class DatasetUtils {
+        preserveCase: boolean,
+        overrideExtension?: string
+    ): Promise<{ [key: string]: string }> {
+        ZoweLogger.trace("dataset.utils.getExtensionMap called.");
```
i know we are not very consistent, but should `trace` log messages match the `class.method` names exactly? 😅 
No need to do this here though 😋

zFernand0 commented on packages/zowe-explorer/package.json:
```diff
@@ -1724,6 +1782,18 @@
+        {
+          "command": "zowe.ds.downloadMember",
+          "when": "never"
+        },
```
I know this is outside of the scope of this PR, but at some point we should consider exposing all of these commands int he command palette 🙏

zFernand0 reviewed (changes requested):
LGTM! (so far) 😋 
---
I have yet to try the functionality, plus i'm missing about 6 files to review 😅 

I may have a second round of comments later today.

The only reason for my requested changes was due to some edge cases arroud recursion 🙏

JWaters02 commented on packages/zowe-explorer-api/src/extend/MainframeInteraction.ts:
```diff
@@ -96,7 +96,7 @@ export namespace MainframeInteraction {
          *     minimal properties name, mode.
          */
-        fileList(ussFilePath: string): Promise<zosfiles.IZosFilesResponse>;
+        fileList(ussFilePath: string, options?: zosfiles.IUSSListOptions): Promise<zosfiles.IZosFilesResponse>;
```
Hmm @JillieBeanSim I do not understand why this would be a breaking change or not backwards compatible, as I thought simply making a parameter optional is enough to get by.

JWaters02 commented:
> Thanks for working on this @JWaters02 i have left a few comments/questions and also with the discussion on the UI I wonder if we should keep compatibility with the jobs tree search without the check boxes and clicking on the item takes them to a way to fill in the value. Any boolean values we need to clearly document the default in description and docs.

Thanks for the review @JillieBeanSim in terms of the jobs search UI part which part of the UI are you specifically talking about? The USS advanced filters part? I think maybe it might make sense there as currently if you select multiple filters to change it brings up each one in a row, but with jobs search design they would change each part in turn. So it would more clicks/actions but would probably make more sense UX wise.

As for the other parts (top level option quick picks) just having the checkboxes makes most sense to me, because they are just booleans and don't need some extra string or number box to specify value.

JWaters02 commented:
As for the documentation comment, I have a WIP PR open in zowe/docs-site for downloads which I will do a pass over once this one is merged into main as some things have changed since I initially wrote that.

adam-wolfe commented:
The UX looks fine to me and the things I've tried have worked as expected.

The only confusing thing I found was that it is not clear what effect `Override File Extension` has when downloading members. If I go to download members from a PDS named `ADAMW.TEST.CBL` the quick pick says "current: jcl", however, regardless of whether the Override File Extension checkbox is checked, the downloaded files will be given a .cbl extension (which is what I would want; but what is the purpose of the checkbox then?)
<img width="1130" height="227" alt="image" src="https://github.com/user-attachments/assets/d545f900-1ef7-4a58-a76f-743f8cb1e79d" />

traeok commented on packages/zowe-explorer/src/trees/dataset/DatasetInit.ts:
```diff
@@ -235,6 +235,20 @@ export class DatasetInit {
+
+        context.subscriptions.push(
+            vscode.commands.registerCommand("zowe.ds.downloadDataSet", async (node: IZoweDatasetTreeNode) => DatasetActions.downloadDataSet(node))
+        );
```
These two (before/after changes) are functionally equivalent:

```typescript
// Before
vscode.commands.registerCommand("zowe.ds.downloadDataSet", async (node: IZoweDatasetTreeNode) => DatasetActions.downloadDataSet(node));
// After
vscode.commands.registerCommand("zowe.ds.downloadDataSet", async (node: IZoweDatasetTreeNode): Promise<void> => {
    await DatasetActions.downloadDataSet(node);
});
```

Command execution logic in VS Code indicates that all promises are awaited from command handlers:

https://github.com/microsoft/vscode/blob/7e715cdc2d29afed47950ed6d7ee66f528480d82/src/vs/workbench/api/common/extHostCommands.ts#L248

```typescript
		const stopWatch = StopWatch.create();
		try {
			return await callback.apply(thisArg, args);
		} catch (err) {
```

JWaters02 replied:
Makes sense, good to have expert opinion! haha!
